diffkemp/semdiff/caching.py

```diff
@@ -0,0 +1,312 @@
+"""
+A mechanism for caching the results of individual SimpLL runs.
+
+The function comparison results that have been collected so far are stored in
+a graph whose vertices are the individual results (i.e. whether two functions
+are equal or not) and the edges are function calls. This graph is represented
+by the ComparisonGraph class.
+"""
+
+from diffkemp.semdiff.result import Result
+from enum import IntEnum
+
+
+class ComparisonGraph:
+    """
+    Representation of a (partial or complete) result of a DiffKemp run
+    in comparison mode.
+    """
+    class Side(IntEnum):
+        """Either the left module or the right module."""
+        LEFT = 0
+        RIGHT = 1
+
+    class DependencyKind(IntEnum):
+        """
+        A strong dependency means that the equality of the target affects
+        the equality of the source. A weak dependency means that the edge (or
+        edges to the vertex, if used with one) is/are important for lookups in
+        the graph, but the non-equality of the target should not generate
+        a diff.
+
+        Note: A normalized graph is a graph where all vertices are strong (i.e.
+        they do not contain dotted functions).
+        """
+        WEAK = 0
+        STRONG = 1
+
+    class Vertex:
+        """
+        Vertex in the comparison graph.
+
+        Represents a single function difference and contains both function
+        names, the comparison result, a list of attached non-function
+        differences and a list of its edges (i.e. the direct callees).
+
+        Note: name, file and line are tuples containing the values for both
+        modules.
```

Comment:
I guess in such case `names`, `files`, and `lines` would be more appropriate names.